[4.0] Changing checked_out_time from nulldate to real null value

Hackwar opened the pull request:
Since MySQL 8 does not support the 0000-00-00 00:00:00 default value anymore, we have to fix this properly and use a real null value instead. I was looking for some way to cut up the work in good chunks and for me it was along the lines of features. So the checkin feature is the first that I've tackled here. This should change all columns and the code to properly use the null value instead. The Postgres update code is untested as of yet.

Hackwar committed:
Changing checked_out_time from nulldate to real null value

richard67 commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-07-27.sql:
```diff
@@ -0,0 +1,34 @@
+ALTER TABLE `#__banner_clients` ALTER `checked_out_time` DROP DEFAULT;
+ALTER TABLE `#__banner_clients` CHANGE COLUMN `checked_out_time` `checked_out_time` DATETIME NULL;
```
```suggestion
ALTER TABLE `#__banner_clients` MODIFY COLUMN `checked_out_time` DATETIME NULL DEFAULT NULL;
```
1. CHANGE is only necessary when the column name shall be changed, too, otherwise we should use MODIFY.
This should be changed for all the following column change statements, too.
2. We should do that in 1 statement, so the database checker also can handle it. So change this line as shown here and remove the previous one with `DROP DEFAULT`. But I am not 100 % sure if that will work on 8.0, too, or if it needs the 2 separate staments. In other PRs, e.g. 25362 or 25360 it was done like I say, but if that fails then go back to separate statements and just change "CHANGE COLUMN `checked_out_time` `checked_out_time`" to "MODIFY COLUMN `checked_out_time`".

richard67 committed:
Corrections for MySQL schema update

Change alter table statements to 1 with MODIFY, correct column name for ucm_content table in UPDATE statement.

Hackwar committed:
Merge pull request #27 from richard67/patch-1

Corrections for MySQL schema update

richard67 committed:
Update 4.0.0-2019-07-27.sql

SQL code style: `null` -> `NULL`

richard67 committed:
More SQL code style null to NULL

richard67 committed:
Fix ucm content for postgresql too

Hackwar committed:
Merge pull request #28 from richard67/patch-1

Fix SQL code style null to NULL, correct column name for ucm_content table also for postgresql

richard67 committed:
Correct MySQL schema update syntax errors

Hackwar committed:
Merge pull request #29 from richard67/patch-1

Correct MySQL schema update syntax errors

wilsonge commented:
We're going to need to add this https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Table/Table.php#L125-L131 to affected tables I believe too

richard67 commented:
@wilsonge Would that mean it is for the complete table? In this case it might really need to extend this PR to other columns or make new ones by table for where it's not done yet.

@Hackwar I've successfully tested new install of 4.0-dev + this PR + changes from my PR #25658 with MySQL driver on MySQL 8 db and also with PostgreSQL. For testing updates I have a modified update package of last nightly build with changes from PR and my PR #25658, but this I haven't tested yet.

To you both: How shall we continue?

Hackwar commented:
I'm very much for doing this in separate PRs. As I said, I'd like to tackle this as one generic feature per PR. Maybe the published_up/published_down stuff next?

When we have all datetime fields processed, we can change that attribute, as you wrote @wilsonge.

richard67 commented:
For testing update, modified update package of last nightly build + this PR + changes from my PR #25658 can be found here: [https://test5.richard-fath.de/Joomla_4.0.0-alpha11-dev-Development-Update_Package__pr-25715.zip](https://test5.richard-fath.de/Joomla_4.0.0-alpha11-dev-Development-Update_Package__pr-25715.zip). Will test later.

HLeithner commented:
Updating a feature like checkin over the complete code base is hard to test and then you have tido the same tests for publishedup/down and so on.

wilsonge commented:
But in that case I think we'll find the null fields do not get updated and in that case i don't think you'll be able to uncheckout an article. This is why things were being done per component rather than per db field :)

Quy commented:
Replaced by #26491 and #26490.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25715">issues.joomla.org/tracker/joomla-cms/25715</a>.</sub>
